### PR TITLE
Add new game: Checkers

### PR DESCRIPTION
**Pull request by Jazeem**

### CODE REVIEW

**Comment by lanctot** [comment]
Thanks!

Seems like the checkers test is causing a seg fault. This usually means there's a bad memory access (out of bounds indexing into an array). 

To find it, I recommend building in debug mode (compiler flags -O0 and possibly -g as well) and run the test using valgrind: https://stackoverflow.com/questions/5134891/how-do-i-use-valgrind-to-find-memory-leaks.

**Comment by lanctot** [comment]
Setting the `BUILD_TYPE` environment variable to `Debug` should have all the flags you need, see here https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/CMakeLists.txt#L43. So you need to rebuild from scratch: (remove the build directory entirely, then set the env var and rerun the build) then run it manually through valgrind.

**Comment by Jazeem** [comment]
Thanks. Let me look into it!

**Comment by Jazeem** [comment]
Hi @lanctot, something I've noticed in clobber. In this line, shouldn't the rank of the TensorView always be 3 instead of kCellStates? 

Maybe I'm missing something.

https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L393

**Comment by lanctot** [comment]
> Hi @lanctot, something I've noticed in clobber. In this line, shouldn't the rank of the TensorView always be 3 instead of kCellStates?

`kCellStates` is always 3: one for empty, one for black, and one for white: https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L33.

We used `kCellStates` here to be explicit that there's one plane per cell state (empty, black, white) and also for consistency with the other games that do it this way.

**Comment by lanctot** [comment]
Thanks for fixing. I have not looked at the code yet. Curious: how did you handle the case of multiple captures?

**Comment by Jazeem** [comment]
https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L393

> We used `kCellStates` here to be explicit that there's one plane per cell state (empty, white, black) and also for consistency with the other games that do it this way.

My doubt was, to do this wouldn't you have replaced `kNumPlayers + 1` with `kCellStates` inside the braces? 

Suppose if I were to add another dimension to the Tensor, a boolean to indicate whether the cell is on the boundary or not. Wouldn't that increase the rank from 3 to 4, although `kCellStates` value is 3?

Maybe I've got this jumbled up and should go through tensor_view.h

**Comment by lanctot** [comment]
Sorry -- you are indeed correct. I had forgotten the use of TensorView, but it refers to the number of dimensions and should not be confused with the cell states. (It just happens to be 3 in this case, so it works.)

This should read TensorView<3> in clobber. (Do you want to submit a separate PR? If not, very easy to change on our side).

Yeah, using Tensorview does make encoding the observation tensors easier and less error-prone. See the one in breakthrough for example: https://github.com/deepmind/open_spiel/blob/0ffd9584c8be560f4e0f659b3ad93933d9078b48/open_spiel/games/breakthrough.cc#L326

**Comment by Jazeem** [comment]
> This should read TensorView<3> in clobber. (Do you want to submit a separate PR? If not, very easy to change on our side).

I can create a separate PR. 

P.S. I was checking the contributor graph and noticed 1 commit against my name, even though when I click on it I can see several more. Curious to know if I'm missing any git configurations to have caused this discrepancy.

**Comment by lanctot** [comment]
> P.S. I was checking the contributor graph and noticed 1 commit against my name, even though when I click on it I can see several more. Curious to know if I'm missing any git configurations to have caused this discrepancy.

Hmm, yeah.. I see that but I'm not really sure what the cause is. 

I think our import/export tool (copybara) squashes all the commits from one PR into a single commit when it gets imported. But there has to be more to it than that, because you also updated games.md in a separate PR. Let's see what it looks like after this next PR and Checkers. Could also be how the views get summarized/aggregated on the github side.

**Comment by Jazeem** [comment]
> But since you've mentioned it. I've noticed an edge case I've missed for multiple jumps. I'll update the code, add test cases and push again.

I've fixed this and updated the code, test case

**Comment by Jazeem** [comment]
Hi @lanctot, Did you get a chance to look at the code?

**Comment by lanctot** [comment]
@tewalds will take a look. Apologies for the delay.

File: open_spiel/games/checkers.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPEN_SPIEL_GAMES_CHECKERS_H_
+#define OPEN_SPIEL_GAMES_CHECKERS_H_
+
+// Implementation of the board game Checkers.
+// https://en.wikipedia.org/wiki/Checkers
+//
```
Comment (lanctot) [inline open_spiel/games/checkers.h:20]: I noticed your implementation has rows and columns as paramters. We usually list them all here (see headers of other games)

File: open_spiel/games/checkers.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+// Some notes about this implementation:
+// - Drawing: 
+//     Game is drawn if no pieces have been removed in 40 moves
```
Comment (lanctot) [inline open_spiel/games/checkers.h:23]: 40 Seems a bit arbitrary-- is this a commonly chosen value for this game? If so maybe mention that in the comments here?

File: open_spiel/games/checkers.h [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+ protected:
+  void DoApplyAction(Action action) override;
+
+ private:
+  int ObservationPlane(CellState state, Player player) const;
+
+  Player current_player_ = 0;  // Player zero (White, 'o') goes first.
+  Player outcome_ = kInvalidPlayer;
+  int multiple_jump_piece_ = 0; // Piece in the board who can do multiple jump. Represented by row * rows_ + column
```
Comment (lanctot) [inline open_spiel/games/checkers.h:99]: Can you say a bit more in the comments here about how this variable is used or why it's needed?

**Review by lanctot — CHANGES_REQUESTED** [review]

File: open_spiel/games/checkers.cc [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+
+void CheckersState::UndoAction(Player player, Action action) {
+  std::vector<int> values =
+      UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+
+  const int start_row = values[0];
+  const int start_column = values[1];
+  const int direction = values[2];
+  const int move_type = values[3];
+  const int captured_piece_type = values[4];
```
Comment (tewalds) [inline open_spiel/games/checkers.cc:506]: Ah, I see, this is why the piece types are in there. Is it better to put more details in the history and undo based on that instead of including it in the action?

File: open_spiel/games/checkers_test.cc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+namespace open_spiel {
+namespace checkers {
+namespace {
+
+namespace testing = open_spiel::testing;
+
+void BasicSerializationTest() {
+  std::shared_ptr<const Game> game = LoadGame("checkers");
+  std::unique_ptr<State> state = game->NewInitialState();
+  std::unique_ptr<State> state2 = game->DeserializeState(state->Serialize());
```
Comment (tewalds) [inline open_spiel/games/checkers_test.cc:29]: It seems worth testing a more complicated state.

File: open_spiel/games/checkers_test.cc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  SPIEL_CHECK_EQ(state->ToString(), state2->ToString());
+}
+
+void BasicCheckersTests() {
+  testing::LoadGameTest("checkers");
```
Comment (tewalds) [inline open_spiel/games/checkers_test.cc:34]: Can you test different sizes as well?

File: open_spiel/games/checkers.cc [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+  std::string result = "";
+  for (int r = 0; r < rows_; r++) {
+    // Ensure the row labels are aligned.
+    if (rows_ - r < 10 && rows_ >= 10) {
+      absl::StrAppend(&result, " ");
+    }
+    absl::StrAppend(&result, RowLabel(rows_, r));
+
+    for (int c = 0; c < columns_; c++) {
+      absl::StrAppend(&result, StateToString(BoardAt(r, c)));
```
Comment (tewalds) [inline open_spiel/games/checkers.cc:402]: Is it worth distinguishing between empty black/white squares, as in one that pieces can be on vs ones that are inaccessible? Maybe with dot vs space?

File: open_spiel/games/checkers.cc [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+      SetBoard(end_row, end_column, CellState::kEmpty);
+      break;
+    case MoveType::kCapture:
+      end_row = start_row + kDirRowOffsets[direction] * 2;
+      end_column = start_column + kDirColumnOffsets[direction] * 2;
+      SetBoard(start_row, start_column, player_piece);
+      SetBoard(end_row, end_column, CellState::kEmpty);
+      CellState captured_piece = OpponentState(PlayerToState(player));
+      SetBoard((start_row + end_row) / 2, (start_column + end_column) / 2, 
+        captured_piece_type == PieceType::kMan ? captured_piece : CrownState(captured_piece));
```
Comment (tewalds) [inline open_spiel/games/checkers.cc:530]: This wrapping is a bit odd.

**Review by tewalds — CHANGES_REQUESTED** [review]
Thanks for submitting this game! It's a great addition!

Feel free to push back on some of the more design type comments/suggestions.

It's also worth taking a look at the Google C++ style guide. You followed most of it, but in particular the 80 char line length limit needs a closer look.

File: open_spiel/games/checkers.h (hunk 3 above)
Comment (Jazeem) [reply open_spiel/games/checkers.h:99]: This variable is used to indicate that the last moved piece can make an additional jump (multiple jumps). In that case LegalActions is filtered to only represent moves available for this piece.

I'm not entirely happy with the implementation since I had to use a global variable, if you have other thoughts, do share them

Comment (lanctot) [reply open_spiel/games/checkers.h:99]: Ok great. But, what I meant was: could you elaborate on that in the comment?

Also, can we please clarify what version of the rules you are using (at the top of the header file)? I wasn't aware of forced multi-captures, e.g. I don't see anything about it mentioned on https://en.wikipedia.org/wiki/Checkers. Maybe this is a common convention used in computer Checkers? If so, please clarify that in the documentation at the top of the file.

Comment (Jazeem) [reply open_spiel/games/checkers.h]: It's mentioned in the Wiki link that
> When capturing an opponent's piece is possible, capturing is mandatory in most official rules.
That's why I added those. About the 40 move rule, I could find some mentions of it online
> http://www.flyordie.com/games/help/checkers/en/games_rules_checkers.html

File: open_spiel/games/checkers.h [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "open_spiel/spiel.h"
+
+namespace open_spiel {
+namespace checkers {
+
+inline constexpr int kNumPlayers = 2;
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.h]: That's true. Removed it.

File: open_spiel/games/checkers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.h]: Have fixed these

[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.h]: Yes. This variable is used to identify a piece with a multiple jump possibility that has completed it's first jump. Hence there'll be only one such piece.

File: open_spiel/games/checkers.cc [hunk 10]
```diff
@@ -0,0 +1,546 @@
+// Copyright 2019 DeepMind Technologies Limited
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Fixed

File: open_spiel/games/checkers.cc [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+// Types of pieces: normal & crowned 
+inline constexpr int kNumPieceType = 2;
+// Number of unique directions each piece can take.
+inline constexpr int kNumDirections = 4;
+
+// Index 0: Direction is diagonally up-left.
+// Index 1: Direction is diagonally up-right.
+// Index 2: Direction is diagonally down-right.
+// Index 3: Direction is diagonally down-left.
+constexpr std::array<int, kNumDirections> kDirRowOffsets = {{-1, -1, 1, 1}};
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:45]: Keeping this as is improves readability IMO

File: open_spiel/games/checkers.cc [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
+#include "open_spiel/spiel_utils.h"
+#include "open_spiel/utils/tensor_view.h"
+
+namespace open_spiel {
+namespace checkers {
+namespace {
+
+// Number of rows with pieces for each player
+inline constexpr int kNumRowsWithPieces = 3;
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Fixed

File: open_spiel/games/checkers.cc [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+}
+
+CellState CrownState(CellState state) {
+  switch (state) {
+    case CellState::kWhite:
+      return CellState::kWhiteCrowned;
+    case CellState::kBlack:
+      return CellState::kBlackCrowned;
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Fixed

File: open_spiel/games/checkers.cc [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+  }
+}
+
+CellState CrownStateIfLastRowReached(int row, CellState state) {
+  if (row == 0 && state == CellState::kWhite) {
+    state = CellState::kWhiteCrowned;
+  }
+  if (row == kDefaultRows - 1 && state == CellState::kBlack) {
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Yes. My bad. Have fixed these instances.

File: open_spiel/games/checkers.cc [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+PieceType StateToPiece(CellState state) {
+  switch (state) {
+    case CellState::kWhite:
+    case CellState::kBlack:
+      return PieceType::kMan;
+    case CellState::kWhiteCrowned:
+    case CellState::kBlackCrowned:
+      return PieceType::kKing;
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Fixed

File: open_spiel/games/checkers.cc [hunk 16]
```diff
@@ -0,0 +1,546 @@
...
+      return CellState::kBlack;
+    default:
+      SpielFatalError(absl::StrCat("Invalid player id ", player));
+  }
+}
+
+std::string StateToString(CellState state) {
+  switch (state) {
+    case CellState::kEmpty:
+      return ".";
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:121]: I thought keeping it in a switch statement would improve readability

File: open_spiel/games/checkers.cc [hunk 17]
```diff
@@ -0,0 +1,546 @@
...
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
+  }
+}
+
+CellState PlayerToState(Player player) {
+  switch (player) {
+    case 0:
+      return CellState::kWhite;
+    case 1:
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:111]: Since there are 2 players and 5 cell states, I thought being explicit would be better

File: open_spiel/games/checkers.cc [hunk 18]
```diff
@@ -0,0 +1,546 @@
...
+    SpielFatalError(absl::StrCat("Unknown state ", str));
+  }
+}
+
+CellState OpponentState(CellState state) {
+  return PlayerToState(1 - StateToPlayer(state));
+}
+
+std::string RowLabel(int rows, int row) {
+  int row_number = 1 + (rows - 1 - row);
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: True. This function was a remnant when I cloned Clobber to write checkers. I've simplified this, I can make the changes in Clobber too if needed.

File: open_spiel/games/checkers.cc [hunk 19]
```diff
@@ -0,0 +1,546 @@
...
+  std::string label = std::to_string(row_number);
+  return label;
+}
+
+std::string ColumnLabel(int column) {
+  std::string label = "";
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:163]: This was also part of the Clobber code. But I think rewriting like this could cause a memory leak, what do you think?

File: open_spiel/games/checkers.cc [hunk 20]
```diff
@@ -0,0 +1,546 @@
...
+}  // namespace
+
+std::ostream& operator<<(std::ostream& stream, const CellState& state) {
+  switch (state) {
+    case CellState::kWhite:
+      return stream << "White";
+    case CellState::kBlack:
+      return stream << "Black";
+    case CellState::kWhiteCrowned:
+      return stream << "WhiteCrowned";
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:176]: Yes, makes sense. Have renamed this.

File: open_spiel/games/checkers.cc [hunk 21]
```diff
@@ -0,0 +1,546 @@
...
+
+  moves_without_capture_ = 0;
+  board_ = std::vector<CellState>(rows_ * columns_, CellState::kEmpty);
+
+  for (int row = rows_ - 1; row >= 0; row--) {
+    for (int column = 0; column < columns_; column++) {
+      if ((row + column) % 2 == 1) {
+        if (row >= 0 && row < kNumRowsWithPieces) {
+          SetBoard(row, column, CellState::kBlack);
+        } else if (row >= (kDefaultRows - kNumRowsWithPieces)) {
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Yes. Fixed this.

File: open_spiel/games/checkers.cc [hunk 22]
```diff
@@ -0,0 +1,546 @@
...
+
+void CheckersState::SetCustomBoard(const std::string board_string) {
+  SPIEL_CHECK_EQ(rows_ * columns_, board_string.length() - 1);
+  current_player_ = board_string[0] - '0';
+  // Create the board from the board string. The characters 'o', '8' are White
+  // (first player) & '+', '*' are Black (second player), and the character '.'
+  // is an Empty cell. Population goes from top left to bottom right.
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      char state_character = board_string[1 + row * columns_ + column];
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:232]: Changed this.

File: open_spiel/games/checkers.cc [hunk 23]
```diff
@@ -0,0 +1,546 @@
...
+          SetBoard(row, column, CellState::kWhite);
+        }
+      }
+    }
+  }
+}
+
+void CheckersState::SetCustomBoard(const std::string board_string) {
+  SPIEL_CHECK_EQ(rows_ * columns_, board_string.length() - 1);
+  current_player_ = board_string[0] - '0';
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:224]: Added this

File: open_spiel/games/checkers.cc [hunk 24]
```diff
@@ -0,0 +1,546 @@
...
+    case CellState::kBlack:
+      return CellState::kBlackCrowned;
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
+  }
+}
+
+CellState CrownStateIfLastRowReached(int row, CellState state) {
+  if (row == 0 && state == CellState::kWhite) {
+    state = CellState::kWhiteCrowned;
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Fixed it

File: open_spiel/games/checkers.cc [hunk 25]
```diff
@@ -0,0 +1,546 @@
...
+      char state_character = board_string[1 + row * columns_ + column];
+      CellState state = StringToState(std::string(1, state_character));
+      SetBoard(row, column, state);
+    }
+  }
+}
+
+void CheckersState::DoApplyAction(Action action) {
+  std::vector<int> values =
+      UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Makes sense. Have added a function and struct

File: open_spiel/games/checkers.cc [hunk 26]
```diff
@@ -0,0 +1,546 @@
...
+        multiple_jump_piece_ = end_row * rows_ + end_column;
+      }
+      break;
+  }
+
+  if (!multiple_jump) {
+    current_player_ = 1 - current_player_; 
+  }
+
+  if (LegalActions().empty()) {
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:322]: I don't think so, but it could be put inside the if block.

File: open_spiel/games/checkers.cc [hunk 27]
```diff
@@ -0,0 +1,546 @@
...
+      end_row = start_row + kDirRowOffsets[direction] * 2;
+      end_column = start_column + kDirColumnOffsets[direction] * 2;
+      SPIEL_CHECK_TRUE(InBounds(end_row, end_column));
+      SPIEL_CHECK_EQ(BoardAt(end_row, end_column), CellState::kEmpty);
+      SetBoard((start_row + end_row) / 2, (start_column + end_column) / 2, CellState::kEmpty);
+      SetBoard(end_row, end_column, CrownStateIfLastRowReached(end_row, BoardAt(start_row, start_column)));
+      SetBoard(start_row, start_column, CellState::kEmpty);
+      moves_without_capture_ = 0;
+
+      // Check if multiple jump is possible
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: That's fair. I've added more comments here

File: open_spiel/games/checkers.cc [hunk 28]
```diff
@@ -0,0 +1,546 @@
...
+      std::vector<Action> moves = LegalActions();
+      std::vector<Action> moves_for_last_moved_piece;
+      for (Action action: moves) {
+        std::vector<int> move = UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+        if (move[0] == end_row && move[1] == end_column && move[3] == MoveType::kCapture) {
+          moves_for_last_moved_piece.push_back(action);
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Great suggestion. Code is simplified here now.

File: open_spiel/games/checkers.cc [hunk 29]
```diff
@@ -0,0 +1,546 @@
...
+        }
+      }    
+      if (moves_for_last_moved_piece.size() > 0) {
+        multiple_jump = true;
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Have removed the flag, so there's only a single source of truth now

File: open_spiel/games/checkers.cc [hunk 30]
```diff
@@ -0,0 +1,546 @@
...
+          int adjacent_row = row + kDirRowOffsets[direction];
+          int adjacent_column = column + kDirColumnOffsets[direction];
+
+          if (InBounds(adjacent_row, adjacent_column)) {
+            CellState adjacent_state = BoardAt(adjacent_row, adjacent_column);
+            CellState opponent_state = OpponentState(current_player_state);
+            CellState opponent_state_crowned = CrownState(opponent_state);
+
+            if (adjacent_state == CellState::kEmpty) {
+              action_values[0] = row;                                 // Initial row value of player piece
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Added it

File: open_spiel/games/checkers.cc [hunk 31]
```diff
@@ -0,0 +1,546 @@
...
+              action_values[1] = column;                              // Initial column value of player piece
+              action_values[2] = direction;                           // Direction of move for player piece
+              action_values[3] = MoveType::kNormal;                   // Type of move
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: To make use of in UndoAction. 

Captured piece: Needed to check if player captured a normal or crowned opponent piece
Player piece: Needed because an Undo move from the last row doesn't always mean that the piece should be uncrowned

File: open_spiel/games/checkers.cc [hunk 32]
```diff
@@ -0,0 +1,546 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // If capture moves are possible, it's mandatory to play them.
+  if (!capture_move_list.empty()) {
+    if (multiple_jump_piece_ > 0) {
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Changed this to !=-1

File: open_spiel/games/checkers.cc [hunk 33]
```diff
@@ -0,0 +1,546 @@
...
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      char state_character = board_string[1 + row * columns_ + column];
+      CellState state = StringToState(std::string(1, state_character));
+      SetBoard(row, column, state);
+    }
+  }
+}
+
+void CheckersState::DoApplyAction(Action action) {
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Added in the header comments

File: open_spiel/games/checkers.cc [hunk 34]
```diff
@@ -0,0 +1,546 @@
...
+  }
+  return move_list;
+}
+
+bool CheckersState::InBounds(int row, int column) const {
+  return (row >= 0 && row < rows_ && column >= 0 && column < columns_);
+}
+
+std::string CheckersState::ToString() const {
+  std::string result = "";
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc:422]: Is this to optimise the code performance?

File: open_spiel/games/checkers.cc [hunk 35]
```diff
@@ -0,0 +1,546 @@
...
+void CheckersState::ObservationTensor(Player player,
+                                     absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  TensorView<3> view(values, {kCellStates, rows_, columns_},
+                               true);
+
+  // Observation Tensor Representation:
+  //   Plane 0: 1's where the current player's pieces are, 0's elsewhere.
```
[the remark this answers is not in the document]
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Crowned pieces are added in another Plane 2 & 3

File: open_spiel/games/checkers.cc (hunk 4 above)
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Great suggestion. Have introduced a TurnHistoryInfo struct to do this. Now actions have reduced to 1/4th.

File: open_spiel/games/checkers_test.cc (hunk 5 above)
Comment (Jazeem) [reply open_spiel/games/checkers_test.cc:29]: Added one more test with a board after a few moves

File: open_spiel/games/checkers_test.cc (hunk 6 above)
Comment (Jazeem) [reply open_spiel/games/checkers_test.cc:44]: Added it

File: open_spiel/games/checkers.cc (hunk 7 above)
Comment (Jazeem) [reply open_spiel/games/checkers.cc:431]: I didn't understand this correctly. Can you elaborate?

File: open_spiel/games/checkers.cc (hunk 8 above)
Comment (Jazeem) [reply open_spiel/games/checkers.cc]: Have introduced a function here to improve readability

**Comment by lanctot** [comment]
Merged in https://github.com/deepmind/open_spiel/commit/808e54ef65c4f9d8597fbecf957ce6ead13fb6a1
